src/main/kotlin/org/jitsi/jibri/service/impl/FileRecordingJibriService.kt

```diff
@@ -88,50 +86,52 @@ data class RecordingMetadata(
  * to a file to be replayed later.
  */
-class FileRecordingJibriService(private val fileRecordingParams: FileRecordingParams) : JibriService() {
+class FileRecordingJibriService(
+    private val fileRecordingParams: FileRecordingParams,
+    private val executor: ScheduledExecutorService,
+    private val jibriSelenium: JibriSelenium = JibriSelenium(executor = executor),
+    private val capturer: Capturer = FfmpegCapturer(),
+    private val processFactory: ProcessFactory = ProcessFactory()
+) : JibriService() {
     /**
      * The [Logger] for this class
      */
     private val logger = Logger.getLogger(this::class.qualifiedName)
-    /**
-     * Used for the selenium interaction
-     */
-    private val jibriSelenium = JibriSelenium(
-        executor = Executors.newSingleThreadScheduledExecutor(NameableThreadFactory("JibriSelenium"))
-    )
-    /**
-     * The [FfmpegCapturer] that will be used to capture media from the call and write it to a file
-     */
-    private val capturer = FfmpegCapturer()
     /**
      * The [Sink] this class will use to model the file on the filesystem
      */
     private var sink: Sink
-    /**
-     * If ffmpeg dies for some reason, we want to restart it.  This [ScheduledExecutorService]
-     * will run the process monitor in a separate thread so it can check that it's running on its own
-     */
-    private val executor: ScheduledExecutorService =
-        Executors.newSingleThreadScheduledExecutor(NameableThreadFactory("FileRecordingJibriService"))
     /**
      * The handle to the scheduled process monitor task, which we use to
      * cancel the task
      */
     private var processMonitorTask: ScheduledFuture<*>? = null
+    /**
+     * The directory in which we'll store recordings for this particular session.  This is a directory that will
+     * be nested within [FileRecordingParams.recordingDirectory].
+     */
+    private val sessionRecordingDirectory =
+        fileRecordingParams.recordingDirectory.resolve(fileRecordingParams.sessionId)
+    /**
+     * The directory to where we'll move recording session data if the finalize script fails
+     */
+    private val failedRecordingsDirectory =
+        fileRecordingParams.recordingDirectory.resolve("failed")
 
     init {
+        logger.info("Writing recording to $sessionRecordingDirectory")
         sink = FileSink(
-            fileRecordingParams.recordingDirectory,
+            sessionRecordingDirectory,
             fileRecordingParams.callParams.callUrlInfo.callName
         )
         jibriSelenium.addStatusHandler(this::publishStatus)
     }
 
     override fun start(): Boolean {
-        if (!createIfDoesNotExist(fileRecordingParams.recordingDirectory, logger)) {
+        if (!createIfDoesNotExist(sessionRecordingDirectory, logger)) {
             return false
         }
-        if (!Files.isWritable(fileRecordingParams.recordingDirectory)) {
+        if (!Files.isWritable(sessionRecordingDirectory)) {
```

Comment:
@bgrozev so i guess the idea here would be to change this to some helper which verified all permissions.  technically we'd need to do this at multiple points (since it could change).  i'm starting to think though that, issues there will fail with a proper stack (and we do clean up correctly), maybe we should just let that be the 'handling' of those scenarios.